core/src/main/scala/org/apache/spark/scheduler/Task.scala

```diff
@@ -106,7 +106,22 @@ private[spark] abstract class Task[T](val stageId: Int, var partitionId: Int) ex
     if (interruptThread && taskThread != null) {
       taskThread.interrupt()
     }
-  }  
+  }
+
+  override def hashCode(): Int = {
+    val state = Seq(stageId, partitionId)
+    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
```

Comment:
Maybe a better way is `(stageId + partitionId) * (stageId + partitionId + 1) / 2 +  partitionId`.
See http://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function